File: ML/python/xiaodai/GBTs_classification_spark_xiaodai.py
#!usr/bin/env python
#coding: utf-8

from pyspark import SparkContext
from pyspark.mllib.tree import GradientBoostedTrees, GradientBoostedTreesModel
from pyspark.mllib.util import MLUtils
from pyspark.mllib.linalg import Vectors, SparseVector
from pyspark.mllib.regression import LabeledPoint


class GBT_lkl(object):
    """
    用spark自带的数据集sample_libsvm_data测试spark mllib的
    梯度提升算法
    """

    def __init__(self, data_path="/data/mllib/spark_example_data_set/sample_libsvm_data.txt"):
        """
        做数据的加载
        """
        pass
        #self.data = MLUtils.loadLibSVMFile(sc, data_path)

    def data_process(self):
        """
        做前期的各种数据处理
        """
        trainData, testData = self.data.randomSplit([0.7, 0.3])
        return (trainData, testData)
    
    @classmethod
    def train_model(cls, trianData, cateFeaInfo={}, iterTimes=3):
        """
        训练模型
        """
        model = GradientBoostedTrees.trainClassifier(trianData, \
            categoricalFeaturesInfo=cateFeaInfo, numIterations=iterTimes)
        return model
    
    @classmethod
    def evaluat_model(cls, model, testData):
        """
        评估模型
        testErr:错误率
        precison:精度(TP/(TP+NP))
        recall:召回率(TP/(TP+NF))
        """
        pred = model.predict(testData.map(lambda x:x.features))
        LAP = testData.map(lambda p:p.label).zip(pred)
        testErr = LAP.filter(lambda (v,p):v!=p).count()/float(LAP.count())
        precison = LAP.filter(lambda (v,p):v==p==1).count()/\
                    float(LAP.filter(lambda (v,p):p==1).count())
        recall = LAP.filter(lambda (v,p):v==p==1).count()/\
                    float(LAP.filter(lambda (v,p):v==1).count())

        return {"testErr":testErr, "precison":precison, \
               "recall":recall}
    
    @classmethod
    def save_model(cls, model, path="$SPARK_HOME/gbts_example"):
        """
        保存模型
        """
        model.save(sc, path)
    
    @classmethod
    def load_model(cls, modelclass, path="$SPARK_HOME/gbts_example"):
        """
        重新载入模型
        """
        return modelclass.load(sc, path)


class GBT_lkl_xiaodai(GBT_lkl):
    """
    用小贷数据测SPARK mllib GBTs
    """
    @classmethod
    def data_process_1(cls, data, col_rm):
        """
        做前期的各种数据处理
        @col_rm:要删除数据的列数
        """
        assert isinstance(col_rm, list), "col_rm must be list"
        pdata = data.map(lambda line: _load_rm(line, col_rm))
        resample_time = pdata.filter(lambda p:p.label==0).count()/\
           float(pdata.filter(lambda p:p.label==1).count())
        pdata = pdata.filter(lambda p:p.label==0).union(pdata.filter(lambda p:p.label==1).sample(True, resample_time))
        trainData, testData = pdata.randomSplit([0.7, 0.3])
        return (trainData, testData)
    
def _load_rm(line, col_rm):
    """
    col_rm:要去除的列(从零开始计数)
    """
    tmp = line.strip().split(",")
    features = [tmp[i] for i in xrange(len(tmp)) if i not in col_rm]
    features = ["0" if ele=="" else ele for ele in features]
    return LabeledPoint(tmp[1], features)



if __name__ == '__main__':
    sc = SparkContext(appName='GBTs_xiaodai')

    data_path = "/data/chengqj/myxiaodai.csv"

    col_rm = [0,1,2,4,25,26] + range(19, 24)

    try:
        gbts = GBT_lkl_xiaodai()  #创建对象

        data = sc.textFile(data_path)  #加载数据

        trainData, testData = GBT_lkl_xiaodai.data_process_1(data, col_rm)  #数据处理并切分
        trainData.cache()
        print trainData.first()    

        model = gbts.train_model(trainData, iterTimes=15) #训练模型
        
        #gbts.save_model(model)    #保存模型

        # model_reloaded = gbts.load_model(GradientBoostedTreesModel, "$SPARK_HOME/gbts_example")

        evaluate_result = gbts.evaluat_model(model, testData)  #评估模型

        print evaluate_result, "@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@"

        #feature = ['27', '1', '5', '5', '0', '3', '3', '0', '3000', '3000', '4', '3000', '1', '3', '3000', '3000', 0, '1', '0', '0', '0', '0', '0', '0', '0', '4', 0, '23', '6', '283000', '3', '94333.33333', 0, 0, 0, '2', 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
       
        # features_te = Vectors.dense(feature)

        # print model_reloaded.predict(features_te), "%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%"

    except Exception,e:
        print "error info^^^^^^^^^^^^^^^^^^^^^^^^^^^^^",e

File: ML/scala/xiaodai/gbts_classification_spark_xiaodai.scala

import org.apache.spark._
import org.apache.spark.rdd.RDD
import  scala.collection.mutable.Set
import  org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.mllib.util.MLUtils
import org.apache.spark.mllib.tree.GradientBoostedTrees
import org.apache.spark.mllib.tree.configuration.BoostingStrategy
import org.apache.spark.mllib.tree.model.GradientBoostedTreesModel

import org.apache.spark.SparkContext
import org.apache.spark.SparkConf
import org.apache.spark.mllib.linalg.{Vectors, SparseVector, Vector}
import Array._
import scala.collection.mutable.ArrayBuffer
import scala.math


//去除所有的离散变量
 var  data:RDD[LabeledPoint]  = sc.textFile("/data/chengqj/myxiaodai.csv",10).mapPartitions{
    iter  => 
    var res = List[LabeledPoint]() 
    while (iter.hasNext){
       var  line =iter.next()
       line=line.trim()
       println(line)
       var fields = line.split(",",-1)
       var length = fields.length;
       var vfields =new  ArrayBuffer[Double]() 
       //vfields += fields(0).substring(0, 3).toDouble
       var rm_col = range(19,24).toBuffer
       rm_col += (0,1,2,4,25,26)  //去除所有离散变量
       rm_col ++= range(28,36) 
       //rm_col += (0,1,4,25,26)  //去离散变量，但保留伯努利变量
       // rm_col ++= range(6, 17)  //去除所有分布不均衡的变量
       // rm_col += (24,26,36,37,38,44,46,48,51,52,54,56) //去除所有分布不均衡的变量

       /*var rm_col = range(1, 12).toBuffer  //只保留(12,13,15,16,25,27)这几个变量
       rm_col += (14,26)
       rm_col ++= range(17,25)
       rm_col ++= range(28, 57)*/

       for(i<- 0 to length-1){
         try{
           if(rm_col.exists(el=>el==i))
               println("")
           else
             vfields+=fields(i).toDouble
         
         }catch{                
              case  e:Exception => vfields+=0L
         }          
       }  
       res .::=(LabeledPoint(fields(1).toDouble, Vectors.dense( vfields.toArray)) )
    }
    res.iterator

}

val resample_time = data.filter(p=>p.label==0).count.toDouble/data.filter(p=>p.label==1).count
data = data.filter(p=>p.label==0).union(data.filter(p=>p.label==1).sample(true, resample_time))


val normal_data = data


var splits = normal_data.randomSplit(Array(0.7, 0.3), seed=1)
var (trainingData, testData) = (splits(0), splits(1))
trainingData.cache 

val boostingStrategy = BoostingStrategy.defaultParams("Classification")
boostingStrategy.numIterations = 3 // Note: Use more iterations in practice.
boostingStrategy.treeStrategy.numClasses = 2
boostingStrategy.treeStrategy.maxDepth = 5
//  Empty categoricalFeaturesInfo indicates all features are continuous.
boostingStrategy.treeStrategy.categoricalFeaturesInfo = Map[Int, Int]()

val model = GradientBoostedTrees.train(trainingData, boostingStrategy)

var labelAndPreds = testData.map { point =>
val prediction = model.predict(point.features)
  (prediction,point.label)
}
var testErr = labelAndPreds.filter(r => r._1 != r._2).count.toDouble / testData.count() 
var precision = labelAndPreds.filter(r=>r._1==r._2&&r._2==1).count.toDouble/(labelAndPreds.filter(r=>r._1==r._2&&r._2==1).count.toDouble+labelAndPreds.filter(r=>r._1==1&&r._2==0).count.toDouble)

var recall = labelAndPreds.filter(r=>r._1==r._2&&r._2==1).count.toDouble/(labelAndPreds.filter(r=>r._1==r._2&&r._2==1).count.toDouble+labelAndPreds.filter(r=>r._1==0&&r._2==1).count.toDouble)

var p_n = labelAndPreds.filter(r=>r._2==1).count.toDouble/labelAndPreds.filter(r=>r._2==0).count
println("Test Error = " + testErr)
println("TP/(TP+FP)"+precision+"################################################")
println("TP/(TP+FN)"+recall+"################################################")

